---
title: "enterobacteriaceae-analysis"
output: html_document
---
##Packages
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
devtools::install_github("brennern/ScrambledTreeBuilder")
library("ScrambledTreeBuilder")
library("GenomicBreaks") |> suppressPackageStartupMessages()
library("pheatmap")
library("plotly")
library("dplyr")
```


##Load Data
```{r}
params <- list()
params$resultsDir <- '/flash/LuscombeU/noa/bacteria/results/genomicbreaks'
yamlFiles <- list.files(params$resultsDir, pattern = "*.yaml", full.names = TRUE)
names(yamlFiles) <- yamlFiles |> basename() |> sub(pat = ".yaml", rep="")
```


##Data Formatting
```{r}
EnterobacteriaDataFrame <- formatStats(yamlFiles)

valuesToBuildTheTree <- "percent_identity_global"
treeMatrix <- makeMatrix(EnterobacteriaDataFrame, valuesToBuildTheTree, 100, 50)
valuesToPlaceOnLabels <- "index_avg_strandRand"
valueMatrix <- makeMatrix(EnterobacteriaDataFrame, valuesToPlaceOnLabels, 1, 0.5)
```


##Percent Identity and Mismatches
```{r}
EnterobacteriaDataFrame$percent_identity_global <- EnterobacteriaDataFrame$aligned_matches_bestpair_Total / EnterobacteriaDataFrame$aligned_length_bestpair_Total * 100

EnterobacteriaDataFrame$percent_mismatches_global <- EnterobacteriaDataFrame$mismatching_aligned_bestpair_Total / EnterobacteriaDataFrame$aligned_length_bestpair_Total * 100

ggplot(EnterobacteriaDataFrame) + geom_point() + aes(percent_identity_global,  percent_mismatches_global)
```


##Percent Identity Global
```{r}
pheatmap::pheatmap(as.matrix(cluster::daisy(treeMatrix)), fontsize_row = 5, fontsize_col = 5)

pheatmap::pheatmap(as.matrix(treeMatrix), sym=T, fontsize_row = 5, fontsize_col = 5)
```


##Fraction of Genome Aligned
```{r}
EnterobacteriaDataFrame$fraction_genome_aligned_target <- EnterobacteriaDataFrame$aligned_width_target_bestpair_Total / EnterobacteriaDataFrame$guessed_target_length_bestpair_Total * 100

EnterobacteriaDataFrame$fraction_genome_aligned_query  <- EnterobacteriaDataFrame$aligned_width_query_bestpair_Total  / EnterobacteriaDataFrame$guessed_query_length_bestpair_Total  * 100

EnterobacteriaDataFrame$fraction_genome_aligned_avg    <- (EnterobacteriaDataFrame$fraction_genome_aligned_target + EnterobacteriaDataFrame$fraction_genome_aligned_query) / 2

ggplot(EnterobacteriaDataFrame) + geom_point() + aes(percent_mismatches_global, fraction_genome_aligned_avg, col = percent_identity_global)

ggplot(EnterobacteriaDataFrame) + geom_point() + aes(percent_identity_global,   fraction_genome_aligned_avg, col = percent_mismatches_global)
```


##Fraction of Genome Chained
```{r}
EnterobacteriaDataFrame$fraction_genome_chained_target <- EnterobacteriaDataFrame$chain_width_target_bestpair_Total / EnterobacteriaDataFrame$guessed_target_length_bestpair_Total * 100
EnterobacteriaDataFrame$fraction_genome_chained_query  <- EnterobacteriaDataFrame$chain_width_query_bestpair_Total / EnterobacteriaDataFrame$guessed_query_length_bestpair_Total  * 100
EnterobacteriaDataFrame$fraction_genome_chained_avg    <- (EnterobacteriaDataFrame$fraction_genome_chained_target + EnterobacteriaDataFrame$fraction_genome_chained_query) / 2

ggplot(EnterobacteriaDataFrame) + geom_point() + aes(percent_mismatches_global, fraction_genome_chained_avg, col = percent_identity_global)
```


##Width of Aligned Regions
```{r}
##Aligned Width 1
ggplot(EnterobacteriaDataFrame) + geom_point() + aes(percent_identity_global, aligned_length_Mean, col = percent_mismatches_global)
ggplot(EnterobacteriaDataFrame) + geom_point() + aes(percent_mismatches_global, aligned_length_Mean, col = percent_identity_global)

##Aligned Width 2
ggplot(EnterobacteriaDataFrame) + geom_point() + aes(percent_identity_global, aligned_width_target_bestpair_Mean / guessed_target_length_bestpair_Total, col = percent_mismatches_global) + scale_y_log10()

ggplot(EnterobacteriaDataFrame) + geom_point() + aes(percent_identity_global, aligned_width_query_bestpair_Mean  / guessed_query_length_bestpair_Total, col = percent_mismatches_global) + scale_y_log10()

ggplot(EnterobacteriaDataFrame) + geom_point() + aes(percent_identity_global, aligned_length_bestpair_Mean  / (guessed_target_length_bestpair_Total + guessed_query_length_bestpair_Total) / 2, col = percent_mismatches_global)  + scale_y_log10()
```


##Length of the Chains
```{r}
ggplot(EnterobacteriaDataFrame) + geom_point() + aes(percent_identity_global, chain_width_target_bestpair_Mean, col = percent_mismatches_global) + scale_y_log10()

ggplot(EnterobacteriaDataFrame) + geom_point() + aes(percent_mismatches_global, chain_width_target_bestpair_Mean, col = percent_identity_global) + scale_y_log10()

ggplot(EnterobacteriaDataFrame) + geom_point() + aes(percent_identity_global, chain_width_target_bestpair_Mean / guessed_target_length_bestpair_Mean, col = percent_mismatches_global) + scale_y_log10()

ggplot(EnterobacteriaDataFrame) + geom_point() + aes(percent_identity_global, chain_width_query_bestpair_Mean  / guessed_query_length_bestpair_Mean, col = percent_mismatches_global) + scale_y_log10()
```


##Indices
```{r}
EnterobacteriaDataFrame$index_avg_synteny <- (EnterobacteriaDataFrame$index_synteny_target_bestpair + EnterobacteriaDataFrame$index_synteny_query_bestpair ) / 2

EnterobacteriaDataFrame$index_avg_correlation <- (EnterobacteriaDataFrame$index_correlation_target_bestpair + EnterobacteriaDataFrame$index_correlation_query_bestpair ) / 2

EnterobacteriaDataFrame$index_avg_GOCvicinity4 <- (EnterobacteriaDataFrame$index_GOCvicinity4_target_bestpair + EnterobacteriaDataFrame$index_GOCvicinity4_query_bestpair ) / 2

EnterobacteriaDataFrame$index_avg_strandRand <- (EnterobacteriaDataFrame$index_strandRand_target_bestpair + EnterobacteriaDataFrame$index_strandRand_query_bestpair ) / 2

EnterobacteriaDataFrame[,grepl("index_avg", colnames(EnterobacteriaDataFrame))] |> pairs()
```


##Percent Identity Vs. Strand Randomisation Index Graph
```{r}
EnterobacteriaDataFrame$lab <- ifelse(EnterobacteriaDataFrame$species1 > EnterobacteriaDataFrame$species2,
       paste(EnterobacteriaDataFrame$species1, EnterobacteriaDataFrame$species2, sep = "\n"),
       paste(EnterobacteriaDataFrame$species2, EnterobacteriaDataFrame$species1, sep = "\n"))

EnterobacteriaDataFrame |>
  #dplyr::filter(index_avg_synteny > 0.25) |>
  dplyr::group_by(lab) |>
  dplyr::summarise(percent_identity_global = mean(percent_identity_global),
                   index_avg_strandRand = mean(index_avg_strandRand),
                   lab = lab) |>
  ggplot() +
    geom_point() +
    aes(percent_identity_global, index_avg_strandRand, label=lab) +
    scale_x_continuous("Percent Identity") +
    scale_y_continuous("Strand Randomisation Index") +
    theme_bw() +
    ggtitle("Enterobacteria: Percent Identity Vs. Strand Randomisation Index") +
    geom_text() -> gg
    plotly::ggplotly(gg)
```


##Similarity Vs. Synteny Index
```{r}
ggplot(EnterobacteriaDataFrame) + theme_bw() +
  aes(percent_identity_global, index_avg_synteny, col = percent_mismatches_global) +
  xlab("Identity between aligned regions (%)") +
  ylab("Synteny index") +
  geom_point()
```


##Similarity Vs. Correlation Index
```{r}
ggplot(EnterobacteriaDataFrame) + theme_bw() +
  aes(percent_identity_global, index_avg_strandRand, col = percent_mismatches_global) +
  xlab("Identity between aligned regions (%)") +
  ylab("Strand randomisation index") +
  geom_point() + geom_smooth()
```


##Gene Order Correlation
```{r}
ggplot(EnterobacteriaDataFrame) +
  aes(percent_identity_global, index_avg_GOCvicinity4) +
  aes(color = species1) +
  xlab("Similarity between aligned regions") +
  ylab("abs(correlation index)") +
  geom_point()
```

